sklearn/metrics/classification.py

```diff
@@ -266,17 +266,25 @@ def confusion_matrix(y_true, y_pred, labels=None, sample_weight=None):
     check_consistent_length(sample_weight, y_true, y_pred)
 
     n_labels = labels.size
-    label_to_ind = dict((y, x) for x, y in enumerate(labels))
-    # convert yt, yp into index
-    y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-    y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
-
-    # intersect y_pred, y_true with labels, eliminate items not in labels
-    ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
-    y_pred = y_pred[ind]
-    y_true = y_true[ind]
-    # also eliminate weights of eliminated items
-    sample_weight = sample_weight[ind]
+    # If labels are not consecitive integers starting from zero, then
+    # yt, yp must be converted into index form
+    need_index_conversion = not (
+        labels.dtype.kind in {'i', 'u', 'b'} and
+        np.all(labels == np.arange(n_labels, dtype=labels.dtype)) and
```

Comment:
I don't think the dtype matters for `==` comparison

Reply:
The idea was that it may be faster to build an array of uint8's than int64's, but I'll buy that that isn't an optimization worth sacrificing readability for, especially considering this optimization is only useful when there are a small number of classes.